### ISSUE TITLE
Question on max_length computation in ray_trainer.py

### ISSUE THREAD

**Issue by JefferyChen453** [OpenBMB/RLPR#12]
Why is `max_length` is added by `self.config.data.max_response_length` again here in function `replace_answer_with_gt_batch()`:
https://github.com/OpenBMB/RLPR/blob/8490fca326c625a1697a09ac1a1476a4f6f76325/verl/trainer/ppo/ray_trainer.py#L2481-L2483

`max_length` is already added by `self.config.data.max_response_length` before `replace_answer_with_gt_batch()` is called:
https://github.com/OpenBMB/RLPR/blob/8490fca326c625a1697a09ac1a1476a4f6f76325/verl/trainer/ppo/ray_trainer.py#L2623-L2639

Is this a mistake?

**Comment by jibo27** [comment]
This is not a mistake. When substituting the model's answer with the reference answer, the resulting text may exceed the original length—or the predefined maximum—if the reference answer is particularly long. For simplicity, we set the new `max_length` to `max_prompt_length + 2 * max_response_length` to accommodate such cases.

**Comment by JefferyChen453** [comment]
> This is not a mistake. When substituting the model's answer with the reference answer, the resulting text may exceed the original length—or the predefined maximum—if the reference answer is particularly long. For simplicity, we set the new `max_length` to `max_prompt_length + 2 * max_response_length` to accommodate such cases.

Thanks for your reply.

But I am still confused that
https://github.com/OpenBMB/RLPR/blob/8490fca326c625a1697a09ac1a1476a4f6f76325/verl/trainer/ppo/ray_trainer.py#L2509-L2511
here, the `max_length` is not assigned with the value of `max_prompt_length + 2 * max_response_length `. Which means the `valid_flag` will still be false for overlong input_ids, thus no valid ground_truth_mask will be generated in such cases?

**Comment by jibo27** [comment]
Yes, you are right.

**Comment by Leon-Francis** [comment]
It's quite weird, since `batch_input_ids` is tokenized within the whole batch, and will be padded to the longest sentence length. If one sample exceeds `max_length` after splicing the GT, the entire batch will be invalid. It seems incorrect.

**Comment by Leon-Francis** [comment]
cc @jibo27 @yiranyyu

**Comment by jibo27** [comment]
Thank you for raising this concern! In the `replace_answer_with_gt_batch()` function, we manually pad the new string using the updated maximum sequence length, calculated as `max_prompt_length + 2 * max_response_length`. This process evaluates each sample individually, marking only the specific overlong samples as invalid while keeping others unaffected. The validity of each sample is tracked using the `valid_flags` list. The presence of an overlong sample does not invalidate the entire batch—only the problematic entries are flagged.

If you spot any other potential issues in the implementation, please let us know—we'd be happy to review them.

**Comment by Leon-Francis** [comment]
> Thank you for raising this concern! In the `replace_answer_with_gt_batch()` function, we manually pad the new string using the updated maximum sequence length, calculated as `max_prompt_length + 2 * max_response_length`. This process evaluates each sample individually, marking only the specific overlong samples as invalid while keeping others unaffected. The validity of each sample is tracked using the `valid_flags` list. The presence of an overlong sample does not invalidate the entire batch—only the problematic entries are flagged.
> 
> If you spot any other potential issues in the implementation, please let us know—we'd be happy to review them.

Hi @jibo27, really appreciate the reply! I understand your logic, but what my concerned is here. 

https://github.com/OpenBMB/RLPR/blob/8490fca326c625a1697a09ac1a1476a4f6f76325/verl/trainer/ppo/ray_trainer.py#L2481-L2483

After batch tokenization of the previously constructed `new_texts`, all samples will uniformly pad to the longest one in the batch by default. However, `new_texts` may contain samples with tokens longer than `max_length`. If this happens, all samples will be padded to a length greater than `max_length`. 

https://github.com/OpenBMB/RLPR/blob/8490fca326c625a1697a09ac1a1476a4f6f76325/verl/trainer/ppo/ray_trainer.py#L2509-L2511

After this step, `valid_flags` of all samples will be set to False. I wonder if I'm missing something here. Thanks in advance.

**Comment by jibo27** [comment]
I understand the issue now. The problem occurs when `new_text` exceed `max_length`, and we need to exclude padding tokens when calculating the string length.

A possible solution would be to modify the second condition at L2510 to:
`torch.sum(batch_input_ids[i] != tokenizer.pad_token_id).item() > max_length`

I'll evaluate this change and update the code accordingly. Thanks for catching this!

**Comment by Leon-Francis** [comment]
> A possible solution would be to modify the second condition at L2510 to: `torch.sum(batch_input_ids[i] != tokenizer.pad_token_id).item() > max_length`

This seems to be a proper approach. But need to remove the extra padding afterwards if we want to maintain that the maximum length of the input_ids is `max_length`. I think a better way might be to not pad to the longest text length when getting `batch_input_data` (this should be possible by just tokenization), since padding will be done uniformly later anyway.

**Comment by jibo27** [comment]
> > A possible solution would be to modify the second condition at L2510 to: `torch.sum(batch_input_ids[i] != tokenizer.pad_token_id).item() > max_length`
> 
> This seems to be a proper approach. But need to remove the extra padding afterwards if we want to maintain that the maximum length of the input_ids is `max_length`. I think a better way might be to not pad to the longest text length when getting `batch_input_data` (this should be possible by just tokenization), since padding will be done uniformly later anyway.

Thanks for the suggestion. Whether a text exceeds `max_length` is determined based on its tokenized IDs. Tokenization is more efficient when done in batches rather than individually. However, to the best of my knowledge, batch-wise tokenization requires padding all texts to match the length of the longest one in the batch. As a result, we may still need to remove the padding afterward, but this additional cost should be affordable.
